### PR TITLE
CLI-23

### PR DESCRIPTION
**Pull request by silvexis**
- Unique roles are now created per reactor configuration so that multiple reactors can connect to a single account without stepping on each other.
- Reactor Bucket is no longer a necessary or user configurable setting

### CODE REVIEW

File: cloudzerocli/constants.py [hunk 1]
```diff
@@ -24,7 +24,8 @@ class AwsProviderConnectionType(Enum):
     ROLE = 2
```
Comment (killthrush) [inline cloudzerocli/constants.py]: +1 for normalizing the name templates.

File: cloudzerocli/features/aws/helpers.py [hunk 2]
```diff
@@ -12,14 +12,14 @@
 import cloudzerocli.api.reactor as reactor_api
 import cloudzerocli.utils.aws as aws_utils
 from cloudzerocli import __client_type__, __version__
-from cloudzerocli.constants import SQS_QUEUE_NAME_TEMPLATE, CLOUDTRAIL_SNS_TOPIC_NAME
-from cloudzerocli.utils.interface import get_choice
-from cloudzerocli.utils.aws import apply_cloudzero_reactor_role_policies, create_role, get_role, \
+from cloudzerocli.constants import SQS_QUEUE_NAME_TEMPLATE, CLOUDTRAIL_SNS_TOPIC_NAME, ROLE_NAME_TEMPLATE
```
Comment (imcgaunn) [inline cloudzerocli/features/aws/helpers.py]: `CLOUDTRAIL_SNS_TOPIC_NAME` is a conspicuously better looking name than `SQS_QUEUE_NAME_TEMPLATE` and `ROLE_NAME_TEMPLATE`.

File: cloudzerocli/features/aws/helpers.py [hunk 3]
```diff
@@ -186,7 +182,7 @@ def get_account_name(self):
         else:
             account_name = None
 
-        return account_name or random_name()
```
Comment (imcgaunn) [inline cloudzerocli/features/aws/helpers.py]: i don't think i've ever seen this `random_name()` behavior before. Out of curiosity, what was this for?

File: cloudzerocli/features/aws/helpers.py [hunk 4]
```diff
@@ -272,20 +268,35 @@ def configuration_details(self):
...
                 "remote_cloudtrail_role": self.remote_cloudtrail_role,
                 "remote_cloudtrail_id": self.remote_cloudtrail_id,
-                "reactor_bucket": self.reactor_bucket,
                 "reactor_client": "{} v{}".format(__client_type__, __version__),
                 "region": self.aws_region_name}
 
     def __setup_cloudzero_reactor_role(self):
-        print(' - Checking CLoudZero Reactor Role')
+        role_name = ROLE_NAME_TEMPLATE.format(self.reactor_id)
+
```
Comment (killthrush) [inline cloudzerocli/features/aws/helpers.py]: typo line 277

File: cloudzerocli/features/aws/helpers.py [hunk 5]
```diff
@@ -301,7 +312,7 @@ def __load_and_select_cloudtrail_to_use(self):
             else:  # or just one so we are going to look at that one then, no choice for you!
                 return self.all_cloudtrail_configs[0]
         else:
-            return None  # you got no trail dude
+            return None  # you got no (happy) trail dude
```
Comment (imcgaunn) [inline cloudzerocli/features/aws/helpers.py]: 🙄

File: cloudzerocli/utils/aws.py [hunk 6]
```diff
@@ -105,82 +86,51 @@ def generate_external_id():
...
             'sts:ExternalId']
         return role["Role"]["Arn"], external_id
     except botocore.exceptions.ClientError:
         return None, None
 
 
-def create_role(aws_session, account_id_to_trust):
+def create_cross_account_role(aws_session, account_id_to_trust, role_name):
     """
     Create the aws access role
```
Comment (imcgaunn) [inline cloudzerocli/utils/aws.py]: this docstring could use an update. Not sure what "Create the aws access role" means.

File: cloudzerocli/features/aws/helpers.py [hunk 7]
```diff
@@ -449,12 +460,12 @@ def __create_queue_and_subscribe_to_sns(self, sns_client, sqs_client, sqs_queue_
                                     Endpoint=self.sqs_arn)
         self.valid_sqs = True
 
-    def __setup_cloudtrail(self, reactor_name):
+    def __setup_cloudtrail(self):
         """
         Setup CloudTrail or get the details of the existing CloudTrail configuration
 
         """
```
Comment (killthrush) [inline cloudzerocli/features/aws/helpers.py]: should line 468 also be using `reactor_id`?

Comment (silvexis) [reply cloudzerocli/features/aws/helpers.py]: oooh good catch

File: cloudzerocli/features/aws/helpers.py (hunk 3 above)
Comment (silvexis) [reply cloudzerocli/features/aws/helpers.py]: the idea was if we couldn't find or were given an account name that we would generate a random one for the user, kinda like how docker generates random names for containers, but in hindsight, it was a bad idea.

File: cloudzerocli/__init__.py [hunk 8]
```diff
@@ -5,4 +5,4 @@
 __author__ = 'CloudZero'
 __email__ = 'support@cloudzero.com'
```
[the remark this answers is not in the document]
Comment (silvexis) [reply cloudzerocli/__init__.py]: oooh oops, right, cough...ahem...let me go do that

[the remark this answers is not in the document]
Comment (silvexis) [reply cloudzerocli/__init__.py]: ah, just realized that is handled in the reactor, yes, that should have been a different ticket/pr

File: cloudzerocli/data/example-aws-configuration.yaml [hunk 9]
```diff
@@ -1,21 +1,17 @@
 # These settings are required
-connected_account_name: 'Example AWS Account Name'
+connected_account_name: 'example-aws-account-name'
 connected_account_id: '012345678912'
```
[the remark this answers is not in the document]
Comment (silvexis) [reply cloudzerocli/data/example-aws-configuration.yaml]: ah just an example. The name only has the requirement that it is unique, not that it's a UUID

**Pull request merged**